### PR TITLE
fix error with partial count returned

### PR DESCRIPTION
**Pull request by jasontho-ms**
# Pull Request Template

## Description

This prevents an issue where a partial aggregate can be returned instead of error 429 in some cases.

## Type of change

Please delete options that are not relevant.

- [ x ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 1]
```diff
@@ -353,15 +353,31 @@ public override async Task<QueryResponseCore> DrainAsync(int maxElements, Cancel
             if (this.FailureResponse != null)
             {
                 this.Stop();
-                return this.FailureResponse.Value;
+
+                try
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: This should never throw. Why wrap it in try/finally?

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 2]
```diff
@@ -218,7 +218,7 @@ internal abstract class CosmosCrossPartitionQueryExecutionContext : CosmosQueryE
         /// <summary>
         /// Gets a value indicating whether the context still has more results.
         /// </summary>
-        private bool HasMoreResults => this.itemProducerForest.Count != 0 && this.CurrentItemProducerTree().HasMoreResults;
+        private bool HasMoreResults => this.FailureResponse != null || (this.itemProducerForest.Count != 0 && this.CurrentItemProducerTree().HasMoreResults);
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]: Please update the changelog to note the fix.

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 3]
```diff
@@ -353,15 +353,31 @@ public override async Task<QueryResponseCore> DrainAsync(int maxElements, Cancel
             if (this.FailureResponse != null)
             {
                 this.Stop();
-                return this.FailureResponse.Value;
+
+                try
+                {
+                    return this.FailureResponse;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]:
```suggestion
                    return this.FailureResponse.Value;
```

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs [hunk 4]
```diff
@@ -353,15 +353,31 @@ public override async Task<QueryResponseCore> DrainAsync(int maxElements, Cancel
...
             // Drain the results. If there is no results and a failure then return the failure.
             IReadOnlyList<CosmosElement> results = await this.InternalDrainAsync(maxElements, cancellationToken);
             if ((results == null || results.Count == 0) && this.FailureResponse != null)
             {
                 this.Stop();
-                return this.FailureResponse.Value;
+
+                try
+                {
+                    return this.FailureResponse;
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosCrossPartitionQueryExecutionContext.cs]:
```suggestion
                   return this.FailureResponse.Value;
```

File: Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs [hunk 5]
```diff
@@ -106,7 +106,7 @@ protected override string ContinuationToken
             // With this information we have captured the progress for all partitions in a single continuation token.
             get
             {
-                if (this.IsDone)
+                if (this.IsDone || this.FailureResponse != null)
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs]: What are CT semantics in-case of errors. Is treating it same as success done right path?

Comment (bchong95) [inline Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs]: Why is this needed? Won't order by inherit IsDone from the parent class which has this check?

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/Core/ExecutionContext/CosmosOrderByItemQueryExecutionContext.cs]: CT semantics for error scenario returns null. This is how all properties are handled even in the point operations.
